internal/net/buffer_test.go

```diff
@@ -0,0 +1,267 @@
+	// Read once.
+	packet := make([]byte, 4)
+	var raddr net.Addr
+	n, raddr, err = buffer.ReadFrom(packet)
+	assert.NoError(err)
+	assert.Equal(4, n)
+	assert.Equal([]byte{0, 1, 2, 3}, packet[:n])
+	assert.Equal(addr, raddr)
+
+	// Write again.
+	n, err = buffer.WriteTo([]byte{9, 10, 11}, addr)
+	assert.NoError(err)
+	assert.Equal(3, n)
+
+	// Verify underlying buffer length.
+	// No change in buffer size.
+	assert.Equal(4, len(buffer.packets))
+
+	// Write again and verify buffer grew.
+	n, err = buffer.WriteTo([]byte{12, 13, 14, 15, 16, 17, 18, 19}, addr)
+	assert.NoError(err)
+	assert.Equal(8, n)
+	assert.Equal(4, len(buffer.packets))
+
+	// Close.
+	assert.NoError(buffer.Close())
+}
+
+func TestBufferAsync(t *testing.T) {
+	assert := assert.New(t)
+
+	buffer := NewPacketBuffer()
+	addr, err := net.ResolveUDPAddr("udp", "127.0.0.1:5684")
+	assert.NoError(err)
+
+	// Start up a goroutine to start a blocking read.
+	done := make(chan struct{})
+	go func() {
+		packet := make([]byte, 4)
+
+		n, raddr, err := buffer.ReadFrom(packet)
+		assert.NoError(err)
+		assert.Equal(2, n)
+		assert.Equal([]byte{0, 1}, packet[:n])
+		assert.Equal(addr, raddr)
+
+		_, _, err = buffer.ReadFrom(packet)
+		assert.Equal(io.EOF, err)
+
+		close(done)
+	}()
+
+	// Wait for the reader to start reading.
+	time.Sleep(time.Millisecond)
+
+	// Write once
+	n, err := buffer.WriteTo([]byte{0, 1}, addr)
+	assert.NoError(err)
+	assert.Equal(2, n)
+
+	// Wait for the reader to start reading again.
+	time.Sleep(time.Millisecond)
+
+	// Close will unblock the reader.
+	assert.NoError(buffer.Close())
+
+	<-done
+}
+
+func TestBufferAlloc(t *testing.T) {
+	packet := make([]byte, 1024)
+	addr, err := net.ResolveUDPAddr("udp", "127.0.0.1:5684")
+	if err != nil {
+		t.Fatalf("net.ResolveUDPAddr: %v", err)
+	}
+
+	test := func(f func(count int) func(), count int, max float64) func(t *testing.T) {
+		return func(t *testing.T) {
+			allocs := testing.AllocsPerRun(3, f(count))
+			if allocs > max {
+				t.Errorf("count=%v, max=%v, got %v",
+					count, max, allocs,
+				)
+			}
+		}
+	}
+
+	w := func(count int) func() {
+		return func() {
+			buffer := NewPacketBuffer()
+			for i := 0; i < count; i++ {
+				_, err := buffer.WriteTo(packet, addr)
+				if err != nil {
+					t.Errorf("WriteTo: %v", err)
+					break
+				}
+			}
+		}
+	}
+
+	t.Run("100 writes", test(w, 100, 27))
+	t.Run("200 writes", test(w, 200, 32))
+	t.Run("400 writes", test(w, 400, 41))
+	t.Run("1000 writes", test(w, 1000, 51))
+
+	wr := func(count int) func() {
+		return func() {
+			buffer := NewPacketBuffer()
+			for i := 0; i < count; i++ {
+				_, err := buffer.WriteTo(packet, addr)
+				if err != nil {
+					t.Fatalf("Write: %v", err)
+				}
+				_, _, err = buffer.ReadFrom(packet)
+				if err != nil {
+					t.Fatalf("ReadFrom: %v", err)
+				}
+			}
+		}
+	}
+
+	t.Run("100 writes and reads", test(wr, 100, 7))
+	t.Run("1000 writes and reads", test(wr, 1000, 7))
+	t.Run("10000 writes and reads", test(wr, 10000, 7))
```

Comment:
When we have perfect read / write match, we see identical allocation counts to the `packetio.Buffer`

Reply:
It is unclear to me why these are failing on the Go 1.19 / 1.20 test runs but not on the corresponding Go i386 1.19 / 1.20 test runs. Furthermore, it appears if I update the values to match the Go 1.19 / 1.20 runs, the tests still pass on the Go i386 runs. The tests pass locally on my `linux/amd64` machine.

Reply:
I am going to drop these for now as I'm not sure this is how we should be measuring allocations anyway. It would likely make more sense to invest in tracking benchmark regressions rather than trying to measure exact number of allocs on a few somewhat unrealistic use cases.